### PR TITLE
[Prototype] Use builder pattern for PipelineOptions

### PR DESCRIPTION
**Pull request by pakrym**
Early prototype to explore using the builder pattern for pipeline configuration.

Advantages:
1. Solves API discoverability problem - all extension methods are on HttpPipelineOptions and instantly visible to the customer
2. Separating registrations from instantiation allows reusing the same pipeline options in multiple clients
3. Allows to plug DI enabled policies easily (pass in a factory that resolves policy from DI)
4. Allows to hide policy implementation classes and expose only extension methods simplifying future support
5. Allows to easily extend pipeline sections list just by adding an entry to the enum

I'll clean it up and add tests if we decide that we are willing to pursue this direction.

### CODE REVIEW

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample6_ConfiguringRetries.cs [hunk 1]
```diff
@@ -17,15 +16,15 @@ public partial class ConfigurationSamples
...
-                retriableCodes: new int[] {
-                    500, // Internal Server Error 
-                    504  // Gateway Timeout
-                }
-            );
+                retriableCodes: new int[]
+                {
+                    500, // Internal Server Error
+                    504 // Gateway Timeout
+                });
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample6_ConfiguringRetries.cs]: Should this use params syntax?

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs [hunk 2]
```diff
@@ -20,29 +20,26 @@ public partial class ConfigurationSamples
...
         {
             // this instance will hold pipeline creation options
-            var options = ConfigurationClient.CreateDefaultPipelineOptions();
+            var options = new HttpPipelineOptions();
 
             // specify custon HttpClient
-            options.Transport = new HttpClientTransport(s_client);
-
-            // remove logging policy
-            options.LoggingPolicy = null;
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: We need a way to remove the default policies that will be added inside client. How do you envision this working?

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs [hunk 3]
```diff
@@ -20,29 +20,26 @@ public partial class ConfigurationSamples
...
             // this instance will hold pipeline creation options
-            var options = ConfigurationClient.CreateDefaultPipelineOptions();
+            var options = new HttpPipelineOptions();
 
             // specify custon HttpClient
-            options.Transport = new HttpClientTransport(s_client);
-
-            // remove logging policy
-            options.LoggingPolicy = null; 
+            options.UseHttpClient(s_client);
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: How will people know that transports are exclusive? i.e. will we also have UseXxx for all other transports?

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs [hunk 4]
```diff
@@ -20,29 +20,26 @@ public partial class ConfigurationSamples
...
                 retriableCodes: new int[] {
-                    500, // Internal Server Error 
+                    500, // Internal Server Error
                     504  // Gateway Timeout
                 }
             );
 
             // add a policy (custom behavior) that executes once per client call
-            options.AddPerCallPolicy(new AddHeaderPolicy());
+            options.Append(HttpPipelineSection.PreRetry, next => new AddHeaderPolicy(next));
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: I really dislike the delegates. Also, I find "section" terminology to be strange. Maybe we should do:
```c#
public enum PolicyRunner {
    OncePerCall,
    OncePerRetry,
}
```
or something like that.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs [hunk 5]
```diff
@@ -52,21 +49,35 @@ public async Task ConfiguringPipeline()
             await client.DeleteAsync("some_key");
         }
 
-        class AddHeaderPolicy : HttpPipelinePolicy
+        class AddHeaderPolicy: HttpPipelinePolicy
         {
-            public override async Task ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+            private readonly HttpPipelinePolicy _next;
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: I really don't think policies should have the next filed. I want policies to be reusable between pipelines (if possible).

File: src/SDKs/Azure.Base/data-plane/Azure.Base.Tests/PipelineTests.cs [hunk 6]
```diff
@@ -18,11 +18,11 @@ public class PipelineTests
...
 
         [Test]
-        public void Basics() {
-
-            var options = new HttpPipelineOptions(new MockTransport(500, 1));
-            options.RetryPolicy = new CustomRetryPolicy();
-            options.LoggingPolicy = new LoggingPolicy();
+        public void Basics()
+        {
+            var options = new HttpPipelineOptions()
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base.Tests/PipelineTests.cs]: I am not a big fan of fluent APIs. Can we start with:
```c#
var options = new HttpPipelineOptions();
options.AddLogging();
options.Append(...);
...    
 
```

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using Azure.Base.Http;
+using Azure.Base.Http.Pipeline;
+
+namespace Azure
+{
+    public static class ClientPipelineBuilderExtensions
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: Why are these extensions and not instance methods on the builder? I think shared/common APIs should be instances, and then some services should have extensions with policies that are specific to the service.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                foreach (var pipelineBuilderPipelinePolicy in section)
+                {
+                    currentPolicy = pipelineBuilderPipelinePolicy.PolicyFactory(currentPolicy);
+                }
+            }
+
+            return new HttpPipeline(transport, currentPolicy);
+        }
+
+        public static HttpPipelineOptions AddLogging(this HttpPipelineOptions options, int[] excludeCodes = null)
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: Why do we need this? The current architectural guidelines say that logging will be on by default.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            return options.Append(HttpPipelineSection.PostRetry, next => new LoggingPolicy(next, excludeCodes));
+        }
+
+        public static HttpPipelineOptions AddFixedRetry(this HttpPipelineOptions options, int maxRetries, TimeSpan delay, params int[] retriableCodes)
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: If people want to change max retries only, how will they know what status codes they should pass?

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+        public static HttpPipelineOptions Replace(this HttpPipelineOptions options, HttpPipelineSection section, Func<HttpPipelinePolicy, HttpPipelinePolicy> policyFactory)
+        {
+            foreach (var registration in options.PipelinePolicies.Where(registration => registration.Section == section).ToArray())
+            {
+                options.PipelinePolicies.Remove(registration);
+            }
+
+            return options.Append(section, policyFactory);
+        }
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: We need the ability to add an arbitrary transport. Also, we need the ability to remove policies added by the service, eg. logging policy. Also, it would be nice if the telemetry policy was something that could be removed.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/FixedPolicy.cs [hunk 11]
```diff
@@ -5,19 +5,17 @@
 
 namespace Azure.Base.Http.Pipeline
 {
-    class FixedPolicy : RetryPolicy {
-        int _maxRetries;
-        TimeSpan _delay;
-        int[] _retriableCodes;
+    public class FixedPolicy : RetryPolicy {
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/FixedPolicy.cs]: FixedPolicy was a good name for an internal type. I think not that you made it public, it should be called FixedRetryPolicy. Having said that, why are you making it public?

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs [hunk 12]
```diff
@@ -15,16 +15,10 @@ namespace Azure.Base.Http.Pipeline
...
+        public HttpClient Client { get; set; }
 
-        readonly HttpClient _client;
-
-        public HttpClientTransport(HttpClient client = null)
-            => _client = client == null ? s_defaultClient : client;
-
-        public readonly static HttpClientTransport Shared = new HttpClientTransport();
-
-        public sealed override HttpMessage CreateMessage(IServiceProvider services, CancellationToken cancellation)
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs]: Why did you remove IServiceProvider?

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/LoggingPolicy.cs [hunk 13]
```diff
@@ -13,25 +13,27 @@ public class LoggingPolicy : HttpPipelinePolicy
...
 
-        int[] _excludeErrors = Array.Empty<int>();
+        private readonly HttpPipelinePolicy _next;
+        private readonly int[] _excludeCodes;
 
-        public readonly static LoggingPolicy Shared = new LoggingPolicy();
-
-        public LoggingPolicy(params int[] excludeErrors)
-            => _excludeErrors = excludeErrors;
+        public LoggingPolicy(HttpPipelinePolicy next, int[] excludeCodes)
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/LoggingPolicy.cs]: should we call these explicitly StatusCodes everywhere? as opposed to shortening to "codes"

File: src/SDKs/Azure.Base/data-plane/Azure.Base/HttpPipelineOptions.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+        public ICollection<PolicyRegistration> PipelinePolicies => _pipelinePolicies;
+
+        public Func<HttpPipelineTransport> TransportPolicy { get; set; }
+
+        public bool DisableTelemetry { get; set; }
+
+        public string ApplicationId { get; set; }
+
+        public HttpPipelineOptions Clone()
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/HttpPipelineOptions.cs]: Should this be threadsafe?

File: src/SDKs/Azure.Base/data-plane/Azure.Base/PolicyRegistration.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Azure.Base.Http.Pipeline;
+
+namespace Azure
+{
+    public struct PolicyRegistration
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.Base/data-plane/Azure.Base/PolicyRegistration.cs]: Why in the root namespace and in general why do we need this?

**Review by KrzysztofCwalina — CHANGES_REQUESTED** [review]
In general, I like the idea of the builder. I think it will improve the current programming model a lot. But I am not thrilled with some of the details:
1. Policies having _next field.
2. Use of delegates for policies
3. Other smaller issues in comments.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/HttpPipelineOptions.cs (hunk 14 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/HttpPipelineOptions.cs]: https://github.com/Azure/azure-sdk-for-net/issues/5503

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs (hunk 12 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs]: Presumably, you need IServiceProvider in case a policy needs to revolve types from DI. But in that case you would prefer the policy type to be resolved from DI itself which new design allows to do (the factory delegate would call into DI) so there is no need to put IServiceProvider on message.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs (hunk 10 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: We need to decide who can override who.

It's either "customer can override the client default" or "client can add default policies to what client passed in" we can't have both, so which one should we pick?

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs (hunk 9 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: How do they do it in the current design? I'm not trying to fix every problem we have here.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs (hunk 5 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: Why?

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs (hunk 3 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: > i.e. will we also have UseXxx for all other transports?

Yes. But what other transports?

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample6_ConfiguringRetries.cs (hunk 1 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample6_ConfiguringRetries.cs]: Sure.

File: src/SDKs/Azure.Base/data-plane/Azure.Base.Tests/PipelineTests.cs (hunk 6 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base.Tests/PipelineTests.cs]: Sure

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs (hunk 2 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: https://github.com/Azure/azure-sdk-for-net/pull/5524#discussion_r267514079

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs [hunk 16]
```diff
@@ -35,11 +35,10 @@ public class ConfigurationMockTests
 
         private static (ConfigurationClient service, TestPool<byte> pool) CreateTestService(MockHttpClientTransport transport)
         {
-            HttpPipelineOptions options = ConfigurationClient.CreateDefaultPipelineOptions();
+            var options = new HttpPipelineOptions();
             var testPool = new TestPool<byte>();
-            options.AddService(testPool, typeof(ArrayPool<byte>));
 
-            options.Transport = transport;
+            options.TransportPolicy = () => transport;
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs]: I wanted to avoid having pipeline authors think about state management inside a policy that might be attached to multiple different pipelines. Makes things very simple, policy object might only be in one stop of a single pipeline.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/LoggingPolicy.cs (hunk 13 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/LoggingPolicy.cs]: Maybe we should but it's outside of the scope of this PR.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs (hunk 7 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: Can be both, personal preference. I would still keep AddPolicy methods off PipelineOptions but Append/Remove/etc can be moved to the instance.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs (hunk 8 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: I think this comes down to client overriding customer or customer overriding client problem. You both want to be able for a customer to override client and for a client to override customer (add default retry/logging) I don't think that having both is reasonable it makes things way too complicated because there is no one source of truth.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs (hunk 4 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/samples/Sample7_ConfiguringPipeline.cs]: Delegates also align us more with HttpMessageHandler if we decide to go that direction.

I added the `Retry` section that allows you to have multiple retry policies and to override all of them at once if you want to.

See https://github.com/Azure/azure-sdk-for-net/pull/5524/files#diff-97159a9a3212cad4cb8778aeb3b026fcR44

**Comment by pakrym** [comment]
> Use of delegates for policies

1. Enables policies managed by DI using DI scopes (singleton/transient) in the future.
2. Simplifies programming model in default case: no need to worry about cross-pipeline state management.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/PolicyRegistration.cs (hunk 15 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/PolicyRegistration.cs]: Doesn't have to be in the root namespace. We need them so TryAppend/Replace methods knew if policies are registered in some parts of the pipeline.

Comment (KrzysztofCwalina) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/PolicyRegistration.cs]: Before the change the builder knew where policies are registered without the need for a new public types. I want to minimize number of public types unless such new types improve user experience, not our internal implementation.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (KrzysztofCwalina) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs]: pipeline authors is much smaller subset of developers than those who will configure it. I think we should optimize for the people who configure the pipeline above people who will develop custom policies. Policies should be stateless and be reusable in pipelines.

[the remark this answers is not in the document]
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs]: People who configure it would have nice extensions methods to use.

[the remark this answers is not in the document]
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs]: > Policies should be stateless and be reusable in pipelines

Is there a reason for this?

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs (hunk 10 above)
Comment (KrzysztofCwalina) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: Client suggests defaults, user can tweak them, client can veto some tweaks to ensure it works properly. For example, client needs to suggest retry policy, the user can tweak it's settings, but if the user removes authentication policy, the client will make sure that despite of this, the policy is present in the pipeline.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs (hunk 12 above)
Comment (KrzysztofCwalina) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs]: We don't have an end-to-end DI solution in the platform. People new up policies, people want to use services without having to change APIs, and many other problems.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/LoggingPolicy.cs (hunk 13 above)
Comment (KrzysztofCwalina) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/LoggingPolicy.cs]: sounds good.

File: src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs (hunk 12 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/HttpClientTransport.cs]: There is no need for IServiceProvider in current implementation at all. Why complicate things?

File: src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs (hunk 10 above)
Comment (pakrym) [reply src/SDKs/Azure.Base/data-plane/Azure.Base/ClientPipelineBuilderExtensions.cs]: What if the user wants to use a completely different retry policy and ignore the client provided one?

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs [hunk 17]
```diff
@@ -211,11 +210,10 @@ public async Task GetBatch()
...
         public void ConfiguringTheClient()
         {
-            var options = ConfigurationClient.CreateDefaultPipelineOptions();
+            var options = new HttpPipelineOptions();
             options.ApplicationId = "test_application";
-            options.AddService(ArrayPool<byte>.Create(1024 * 1024 * 4, maxArraysPerBucket: 4), typeof(ArrayPool<byte>));
-            options.Transport = new GetMockTransport(s_testSetting.Key, default, s_testSetting);
-            options.RetryPolicy = RetryPolicy.CreateFixed(5, TimeSpan.FromMilliseconds(100), 404);
+            options.TransportPolicy = () => new GetMockTransport(s_testSetting.Key, default, s_testSetting);
+            options.AddFixedRetry(5, TimeSpan.FromMilliseconds(100), 404);
```
[the remark this answers is not in the document]
Comment (bterlson) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/ConfigurationMockTests.cs]: Cosmos seems to have extremely complex retry logic, fwiw (involving fail-overs and what-have-you): https://github.com/Azure/azure-cosmos-js/tree/c33b3fbb45ea32e3e4e60922196619831b9dc5ea/src/retry. I was talking to @southpolesteve about this and he was even doubting their logic could exist in a pipeline due in part to this complexity.

**Comment by pakrym** [comment]
Closing, this served its purpose.
